diesel/src/sqlite/connection/raw.rs

```diff
@@ -292,99 +343,155 @@ extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A
             }
         }
+    });
+
+    let result = std::panic::catch_unwind(|| A::finalize(aggregator))
+        .map(process_sql_function_result::<RetSqlType, Ret>);
+
+    match result {
+        Ok(Ok(value)) => value.result_of(ctx),
+        Ok(Err(e)) => {
+            let msg = e.to_string();
+            unsafe { context_error_str(ctx, &msg) };
+        }
+        Err(_) => {
+            let msg = format!("{}::finalize() panicked", std::any::type_name::<A>());
+            unsafe { context_error_str(ctx, &msg) };
+        }
     }
 }
 
 unsafe fn null_aggregate_context_error(ctx: *mut ffi::sqlite3_context) {
     static NULL_AG_CTX_ERR: &str = "An unknown error occurred. sqlite3_aggregate_context returned a null pointer. This should never happen.";
 
-    ffi::sqlite3_result_error(
-        ctx,
-        NULL_AG_CTX_ERR.as_ptr() as *const _ as *const _,
-        NULL_AG_CTX_ERR.len() as _,
-    );
+    context_error_str(ctx, NULL_AG_CTX_ERR)
+}
+
+unsafe fn context_error_str(ctx: *mut ffi::sqlite3_context, error: &str) {
+    ffi::sqlite3_result_error(ctx, error.as_ptr() as *const _, error.len() as _);
 }
 
-extern "C" fn destroy_boxed_fn<F>(data: *mut libc::c_void)
+#[allow(warnings)]
+extern "C" fn run_collation_function<F>(
+    user_ptr: *mut libc::c_void,
+    lhs_len: libc::c_int,
+    lhs_ptr: *const libc::c_void,
+    rhs_len: libc::c_int,
+    rhs_ptr: *const libc::c_void,
+) -> libc::c_int
 where
-    F: FnMut(&RawConnection, &[*mut ffi::sqlite3_value]) -> QueryResult<SerializedValue>
-        + Send
-        + 'static,
+    F: Fn(&str, &str) -> std::cmp::Ordering + Send + 'static,
 {
+    let user_ptr = user_ptr as *const F;
+    let f = unsafe { user_ptr.as_ref() }.unwrap_or_else(|| {
+        assert_fail!(
+            "An unknown error occurred. user_ptr is a null pointer. This should never happen."
+        );
+    });
+
+    for (ptr, len, side) in &[(rhs_ptr, rhs_len, "rhs"), (lhs_ptr, lhs_len, "lhs")] {
+        if *len < 0 {
+            assert_fail!(
+                "An unknown error occurred. {}_len is negative. This should never happen.",
+                side
+            );
+        }
+        if ptr.is_null() {
+            assert_fail!(
+                "An unknown error occurred. {}_ptr is a null pointer. This should never happen.",
+                side
+            );
+        }
+    }
+
+    let (rhs, lhs) = unsafe {
+        // Depending on the eTextRep-parameter to sqlite3_create_collation_v2() the strings can
+        // have various encodings. register_collation_function() always selects SQLITE_UTF8, so the
+        // pointers point to valid UTF-8 strings (assuming correct behavior of libsqlite3).
+        (
+            str::from_utf8_unchecked(slice::from_raw_parts(rhs_ptr as *const u8, rhs_len as _)),
+            str::from_utf8_unchecked(slice::from_raw_parts(lhs_ptr as *const u8, lhs_len as _)),
+        )
+    };
+
+    // It doesn't matter if f is UnwindSafe, since we abort on panic.
```

Comment:
I think similar to the sql function handler we should just return an error here, instead of abort the program.